pkg/lib/repo/local/pull.go

```diff
@@ -0,0 +1,244 @@
+		}
+	}()
+
+	verifier := desc.Digest.Verifier()
+	var offset int64 = 0
+	if stat, err := ingestFile.Stat(); err != nil {
+		return fmt.Errorf("failed to stat ingest file: %w", err)
+	} else if stat.Size() != 0 {
+		p.Debugf("Resuming download for digest %s", desc.Digest.String())
+		numBytes, err := io.Copy(verifier, ingestFile)
+		if err != nil {
+			return fmt.Errorf("failed to resume download: %w", err)
+		}
+		p.Logf(output.LogLevelTrace, "Updating offset to %d bytes", numBytes)
+		offset = numBytes
+	}
+	if _, err := blob.Seek(offset, io.SeekStart); err != nil {
+		return fmt.Errorf("failed to seek in remote resource: %w", err)
+	}
+
+	pwriter := p.ProxyWriter(ingestFile, desc.Digest.Encoded(), desc.Size, offset)
+	mw := io.MultiWriter(pwriter, verifier)
+	if _, err := io.Copy(mw, blob); err != nil {
+		return fmt.Errorf("failed to write file: %w", err)
+	}
+	if !verifier.Verified() {
+		return fmt.Errorf("downloaded file hash does not match descriptor")
+	}
+	if err := ingestFile.Close(); err != nil {
+		return fmt.Errorf("failed to close temporary ingest file: %w", err)
+	}
+	blobPath := l.BlobPath(desc)
+	if err := os.Rename(ingestFilename, blobPath); err != nil {
+		return fmt.Errorf("failed to move downloaded file into storage: %w", err)
+	}
+
+	return nil
+}
+
+func (l *localRepo) downloadFile(desc ocispec.Descriptor, blob io.ReadCloser, p *output.PullProgress) (ingestErr error) {
+	ingestDir := constants.IngestPath(l.storagePath)
+	ingestFile, err := os.CreateTemp(ingestDir, desc.Digest.Encoded()+"_*")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary ingest file: %w", err)
+	}
+
+	ingestFilename := ingestFile.Name()
+	// If we return an error anywhere after this point, we want to delete the ingest file we're
+	// working on, since it will never be reused.
+	defer func() {
+		if err := ingestFile.Close(); err != nil && !errors.Is(err, fs.ErrClosed) {
+			p.Logf(output.LogLevelError, "Error closing temporary ingest file: %s", err)
+		}
+		if ingestErr != nil {
+			os.Remove(ingestFilename)
+		}
+	}()
+
+	verifier := desc.Digest.Verifier()
+	pwriter := p.ProxyWriter(ingestFile, desc.Digest.Encoded(), desc.Size, 0)
+	mw := io.MultiWriter(pwriter, verifier)
+	if _, err := io.Copy(mw, blob); err != nil {
+		return fmt.Errorf("failed to write file: %w", err)
+	}
+	if !verifier.Verified() {
+		return fmt.Errorf("downloaded file hash does not match descriptor")
+	}
+	if err := ingestFile.Close(); err != nil {
+		return fmt.Errorf("failed to close temporary ingest file: %w", err)
+	}
+
+	blobPath := l.BlobPath(desc)
+	if err := os.Rename(ingestFilename, blobPath); err != nil {
+		return fmt.Errorf("failed to move downloaded file into storage: %w", err)
+	}
+	return nil
+}
+
+func (l *localRepo) ensurePullDirs() error {
+	blobsPath := filepath.Join(l.storagePath, ocispec.ImageBlobsDir, "sha256")
+	if err := os.MkdirAll(blobsPath, 0777); err != nil {
```

Comment:
I think it would, this is just copying what [oras-go does](https://github.com/oras-project/oras-go/blob/c8c2048e04ce33871adb166d23ee154d70b9a54b/content/oci/storage.go#L166-L169).